opengovsg/plumber
# PLU-5: Email notification for errors

TODO:
- [x] Send email once for errors per pipe (resets at the end of the day)

## Problem
We need to manually email user when there are errors with their pipes.
Improve the experience of users (so they don't get spammed for the same error)

## Solution
- To identify who to send to, fetch user for the flow.
- Create a helper method: `sendErrorEmail` to generate the information to send to the user.
- Automate the process by sending an email in the codebase when the error is thrown for both failed triggers and actions for non test runs.
- To prevent user from getting spammed, store the pipe id as a redis key with some "metadata" which expires at the end of the day.


## Screenshots
Example of email sent:
<img width="1097" alt="image" src="https://github.com/opengovsg/plumber/assets/65110268/aa4c2f4d-984d-4f97-bae9-e6595f72c35e">


## Tests
- Check that only non test run failed executions would cause email to be sent.
- Check that failed executions for the same pipe won't trigger any new emails after the first one (for the day itself)
- Check that it works for different pipes
